Add rgb_range percentile support

atanas-balevsky opened the pull request:
Add support for using percentiles instead of absolute values for the `/rgb`'s endpoint `*_range` parameters.

gives us the possibility to request the `/rgb?r_range=p2,p97...` instead of calling `/metadata` first to fetch the pct's absolute values in order to construct an `/rgb` url

dionhaefner commented on terracotta/server/rgb.py:
```diff
@@ -29,25 +29,34 @@ class Meta:
     b = fields.String(required=True, description="Key value for blue band")
     r_range = fields.List(
-        fields.Number(allow_none=True),
+        fields.String(allow_none=True, validate=validate.Regexp("^p?(\d*\.)?\d+$")),
```
This is not the correct format.

Previously, this accepted any number, which includes scientific notation like `1.3E-2`. Also, percentiles must be ints between 0 and 100 and cannot be float numbers.

To prevent regressions I would advocate for still using `fields.Number` under the hood, with some additional logic to detect the percentile case. A [custom field](https://marshmallow.readthedocs.io/en/stable/custom_fields.html) class should do, similar to [what is described here](https://stackoverflow.com/questions/61614546/python-marshmallow-field-can-be-two-different-types).

dionhaefner commented on tests/handlers/test_rgb.py:
```diff
@@ -74,7 +74,10 @@ def test_rgb_lowzoom(use_testdb, raster_file, raster_file_xyz_lowzoom):
-    "stretch_range", [[0, 20000], [10000, 20000], [-50000, 50000], [100, 100]]
+    "stretch_range", [
+        [0, 20000], [10000, 20000], [-50000, 50000], [100, 100],
+        ["0", "20000"], ["10000", "20000"], ["-50000", "50000"], ["100", "100"],
```
What does this do?

dionhaefner reviewed (commented):
Thanks for the PR.

There are some issues with it that look fixable. More importantly, I want at least one additional contributor to weigh in whether this is a good idea or not.

The interface is getting pretty overloaded, as evidenced by the type annotations:

```python
ListOfRanges = Sequence[Optional[Tuple[Optional[NumberOrString], Optional[NumberOrString]]]]
```

Does anyone else have mixed feelings about this?

atanas-balevsky commented:
Dear @dionhaefner ,
Thanks for the feedback! i've addressed some aspects of it.

now it's much clear that there can be a string or number. The regex is gone and there's more natural way of inferring the range type. i also added a test for the scientific notation when doing absolute values (didn't know you're supporting these)
i'm not sure why the tests action is not running now, but i believe it should be green

mrpgraae commented:
> Thanks for the PR.
> 
> There are some issues with it that look fixable. More importantly, I want at least one additional contributor to weigh in whether this is a good idea or not.
> 
> The interface is getting pretty overloaded, as evidenced by the type annotations:
> 
> ```python
> ListOfRanges = Sequence[Optional[Tuple[Optional[NumberOrString], Optional[NumberOrString]]]]
> ```
> 
> Does anyone else have mixed feelings about this?

I may be biased since @atanas-balevsky is a colleague of mine, and we need this feature for something we are building over at SkyFi. Still, I think this is a good example of leveraging the fact that we have pre-computed information about an image to provide "magic" features like this, so I'd like to convince you that I think it's a good idea, even without my bias 🙂

Regarding the type, I think it was already quite complicated before. The only difference in this PR is that `Number` changed to `NumberOrString`. But if you think it would be better to do the percentiles as a separate query parameter, to not overload this parameter so much, then I don't see anything wrong with that. It would also help make the type validation simpler.

dionhaefner commented on terracotta/server/rgb.py:
```diff
@@ -21,6 +22,36 @@ class RGBQuerySchema(Schema):
+        try:
+            float(data)
+        except ValueError:
+            raise ValidationError("Must be a number")
```
Can we use the same exception for both branches that contains the expected format?

dionhaefner commented on terracotta/handlers/rgb.py:
```diff
@@ -105,3 +106,20 @@ def get_band_future(band_key: str) -> Future:
+        # can be a number
+        return float(scale)
+    raise exceptions.InvalidArgumentsError(
+        "Invalid scale value: %s" % scale
```
Use f-string.

dionhaefner commented on terracotta/handlers/rgb.py:
```diff
@@ -105,3 +106,20 @@ def get_band_future(band_key: str) -> Future:
+            percentile = int(scale[1:]) - 1
+            return metadata["percentiles"][percentile]
+
+        # can be a number
+        return float(scale)
```
Can it though? I don't think passing a string should be valid if not using the percentile feature.

dionhaefner reviewed (changes requested):
Alright, I'm willing to merge this if we can address the issues above, plus:

1. Add full type annotations to all functions + arguments.
2. Ensure pre-commit hooks pass.
3. Add the same argument to `/singleband`.
4. Push coverage to the target.

Thanks @atanas-balevsky.

atanas-balevsky commented on terracotta/server/rgb.py:
```diff
@@ -21,6 +22,36 @@ class RGBQuerySchema(Schema):
+def validate_range(data):
+    if isinstance(data, str) and data.startswith("p"):
+        if not re.match("^p\d+$", data):
+            raise ValidationError("Percentile format is `p<digits>`")
```
Hey @dionhaefner here we'd do only validation of the format. if the number is in range we'd see when we try to access the metadata percentiles: https://github.com/DHI-GRAS/terracotta/pull/322/files#diff-f3ee2be066aa9a6d1432c4e13259b840360f0bf46320bf3b5d8cc07015cf9d09R198-R203 (may save in some cases where metadata[percentiles] is not 0..100

atanas-balevsky commented:
> Alright, I'm willing to merge this if we can address the issues above, plus:
> 
> 1. Add full type annotations to all functions + arguments.
> 2. Ensure pre-commit hooks pass.
> 3. Add the same argument to `/singleband`.
> 4. Push coverage to the target.
> 
> Thanks @atanas-balevsky.

Dear @dionhaefner, 
thanks for the feedback, i believe i addressed all the points. can we rerun the tests & reports?

Happy holidays!
Nas

dionhaefner commented on terracotta/server/fields.py:
```diff
@@ -0,0 +1,37 @@
+from typing import Any, Union
+
+
+class StringOrNumber(fields.Field):
```
Please add a docstring

dionhaefner commented on terracotta/server/fields.py:
```diff
@@ -0,0 +1,37 @@
+            raise ValidationError("Must be a string or a number")
+
+
+def validate_stretch_range(data: Any) -> None:
```
Please add a docstring

dionhaefner commented on terracotta/server/rgb.py:
```diff
@@ -29,25 +30,43 @@ class Meta:
+        description=(
+            "Stretch range [min, max] to use for the red band as JSON array. "
+            "Min and max may be numbers to use as absolute range, or strings "
+            "of the format p<int> with an integer between 0 and 100 to use "
```
Don't use `<>`, angled brackets get parsed as HTML in the API docs. Either escape with `&lt;` and `&gt;` or use a different character.

dionhaefner reviewed (changes requested):
Thanks, this looks really good! Some final nits.

atanas-balevsky commented on tests/server/test_fields.py:
```diff
@@ -0,0 +1,42 @@
+    "args, expected",
+    [("[0, 1]", [0, 1]), ('["p2", "p28"]', ["p2", "p28"]), (None, None)],
+)
+def test_serde(args, expected):
```
serialization/deserialization, had to test the deserialization part of the custom marshmallow types to fit the codecov

atanas-balevsky commented:
Dear @dionhaefner!
Thanks for the feedback. i believe i addressed most of the comments. can we rerun the test actions?

dionhaefner reviewed (approved):
LGTM, thanks!
